File: Compiler/JackCompiler.py
"""
Jack Language Compiler - Main Module.

This is the main entry point for the Jack language compiler. It provides
a command-line interface for compiling Jack source files (.jack) into
Virtual Machine bytecode (.vm) files.

The compiler can process either individual Jack files or entire directories
containing multiple Jack files. It orchestrates the compilation process
by coordinating the tokenizer, compilation engine, and VM writer.

Usage:
    python3 JackCompiler.py <input_file_or_directory>
    
Examples:
    python3 JackCompiler.py Main.jack          # Compile single file
    python3 JackCompiler.py Square/            # Compile all .jack files in directory
"""

from JackTokenizer import JackTokenizer
from CompilationEngine import CompilationEngine
import Config
import os
import sys


def compile_jack_file(input_path: str, output_path: str) -> None:
    """
    Compile a single Jack source file to VM bytecode.
    
    This function handles the complete compilation pipeline for a single
    Jack file: reading the source, tokenizing, parsing, and generating
    VM code.
    
    Args:
        input_path (str): Path to the input .jack file
        output_path (str): Path where the output .vm file should be written
        
    Raises:
        FileNotFoundError: If the input file doesn't exist
        IOError: If there are issues reading/writing files
        SyntaxError: If the Jack source contains syntax errors
    """
    try:
        with open(input_path, "r") as input_file, open(output_path, "w") as output_file:
            # Read the entire source file
            input_file_text = input_file.read()
            
            # Create tokenizer and parse the source
            jack_tokenizer = JackTokenizer(input_file_text)
            
            # Create compilation engine with tokenizer and output file
            compilation_engine = CompilationEngine(output_file, jack_tokenizer)
            
            # Start compilation by advancing to first token and compiling the class
            jack_tokenizer.advance()
            compilation_engine.compileClass()
            
    except FileNotFoundError:
        print(f"Error: Input file '{input_path}' not found")
        sys.exit(1)
    except IOError as e:
        print(f"Error: IO error processing '{input_path}': {e}")
        sys.exit(1)
    except SyntaxError as e:
        print(f"Error: Syntax error in '{input_path}': {e}")
        sys.exit(1)


def compile_directory(directory_path: str) -> None:
    """
    Compile all Jack files in a directory.
    
    Iterates through all .jack files in the specified directory and
    compiles each one to a corresponding .vm file in the same directory.
    
    Args:
        directory_path (str): Path to directory containing .jack files
        
    Note:
        Output .vm files are created in the same directory as the input
        .jack files, with the same base name but .vm extension.
    """
    try:
        for filename in os.listdir(directory_path):
            if filename.endswith(".jack"):
                # Construct input and output file paths
                input_file_path = os.path.join(directory_path, filename)
                output_filename = filename.replace(".jack", ".vm")
                output_file_path = os.path.join(directory_path, output_filename)
                
                print(f"Compiling {filename} -> {output_filename}")
                
                # Compile the individual file
                compile_jack_file(input_file_path, output_file_path)
                
        print(f"Directory compilation completed successfully")
        
    except OSError as e:
        print(f"Error: Failed to process directory '{directory_path}': {e}")
        sys.exit(1)


def main() -> None:
    """
    Main entry point for the Jack compiler.
    
    Parses command line arguments and initiates compilation of either
    a single Jack file or all Jack files in a directory.
    
    Command line usage:
        python3 JackCompiler.py <input_file_or_directory>
        
    The compiler automatically detects whether the input is a file or
    directory and processes accordingly.
    """
    # Check command line arguments
    if len(sys.argv) != 2 or sys.argv[1] in ["-h", "--help"]:
        print("Jack Language Compiler")
        print("======================")
        print("")
        print("Usage: python3 JackCompiler.py <input_file_or_directory>")
        print("")
        print("Arguments:")
        print("  input_file_or_directory  Path to .jack file or directory containing .jack files")
        print("")
        print("Examples:")
        print("  python3 JackCompiler.py Main.jack          # Compile single file")
        print("  python3 JackCompiler.py Square/            # Compile all .jack files in directory")
        print("")
        print("Options:")
        print("  -h, --help              Show this help message")
        print("")
        print("The compiler generates .vm files in the same location as the input .jack files.")
        sys.exit(1)
    
    file_path = sys.argv[1]
    
    # Determine if input is a directory or file
    if os.path.isdir(file_path):
        print(f"Compiling all .jack files in directory: {file_path}")
        compile_directory(file_path)
        
    elif os.path.isfile(file_path):
        # Validate file extension
        if not file_path.endswith(".jack"):
            print(f"Error: '{file_path}' is not a valid .jack file")
            print("Please provide a file with .jack extension or a directory containing .jack files")
            sys.exit(1)
        
        print(f"Compiling single file: {file_path}")
        
        # Generate output filename by replacing .jack with .vm
        output_file_path = file_path.replace(".jack", ".vm")
        
        # Compile the single file
        compile_jack_file(file_path, output_file_path)
        
        print(f"Compilation completed: {output_file_path}")
        
    else:
        print(f"Error: '{file_path}' is not a valid file or directory")
        print("Please check the path and try again")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: CCompiler/JackCompiler.c
/**
 * @file JackCompiler.c
 * @brief Main entry point for the Jack Compiler
 * 
 * This file contains the main entry point for the Jack Compiler (Project 11 from Nand2Tetris),
 * which translates Jack source code (.jack files) into Virtual Machine code (.vm files).
 * The compiler implements a recursive descent parser that translates Jack syntax
 * into VM commands for arithmetic, memory access, control flow, and function calls.
 * 
 * The compiler generates VM code that can be executed by the Virtual Machine
 * emulator, implementing the complete Jack language specification including
 * classes, methods, functions, constructors, variables, expressions, and statements.
 */

#include "Config.h"
#include "CompilationEngine.h"

/**
 * @brief Main entry point for the Jack Compiler
 * 
 * Processes command line arguments and orchestrates the compilation process.
 * Supports both single file compilation and directory processing. For directories,
 * all .jack files are compiled to corresponding .vm files.
 * 
 * @param argc Number of command line arguments
 * @param argv Array of command line argument strings
 * @return 0 on successful compilation, 1 on error
 */
int main(int argc, char * argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <input.jack or directory>\n", argv[0]);
        return 1;
    }
    
    char * fileName = argv[1];
    struct stat pathStat;
    
    if (stat(fileName, &pathStat) == 0) {
        if (S_ISDIR(pathStat.st_mode)) {
            DIR * dir = opendir(fileName);
            if (dir == NULL) {
                fprintf(stderr, "Error: Failed to open directory\n");
                return 1;
            }

            struct dirent * entry;
            while ((entry = readdir(dir)) != NULL) {
                char * extension = strrchr(entry->d_name, '.');
                if (extension != NULL && strcmp(extension, ".jack") == 0) {
                    char fullPath[strlen(fileName) + 1 + strlen(entry->d_name) + 1];
                    snprintf(fullPath, sizeof(fullPath), "%s/%s", fileName, entry->d_name);

                    FILE * inputFile = fopen(fullPath, "r");
                    if (inputFile == NULL) {
                        fprintf(stderr, "Error: Failed to open input file %s\n", fullPath);
                        closedir(dir);
                        return 1;
                    }

                    char outputFileName[strlen(fullPath)];
                    snprintf(outputFileName, sizeof(outputFileName), "%.*s.vm", (int)(strlen(fullPath) - 5), fullPath);

                    FILE * outputFile = fopen(outputFileName, "w");
                    if (outputFile == NULL) {
                        fprintf(stderr, "Error: Failed to open output file %s\n", outputFileName);
                        fclose(inputFile);
                        closedir(dir);
                        return 1;
                    }
                    
                    CompilationEngine * engine = initCompilationEngine(inputFile, outputFile);
                    if (engine == NULL) {
                        fprintf(stderr, "Error: Failed to initialize compilation engine for %s\n", fullPath);
                        fclose(inputFile);
                        fclose(outputFile);
                        closedir(dir);
                        return 1;
                    }
                    
                    compileClass(engine);
                    cleanupCompilationEngine(engine);
                    fclose(inputFile);
                    fclose(outputFile);
                }
            }
            
            closedir(dir);
            return 0;
        } else if (S_ISREG(pathStat.st_mode)) {
            char * extension = strrchr(fileName, '.');
            if (extension == NULL || strcmp(extension, ".jack") != 0) {
                fprintf(stderr, "Error: Invalid file type\n");
                return 1;
            }

            FILE * inputFile = fopen(fileName, "r");
            if (inputFile == NULL) {
                fprintf(stderr, "Error: Failed to open input file\n");
                return 1;
            }

            char outputFileName[strlen(fileName)];
            snprintf(outputFileName, sizeof(outputFileName), "%.*s.vm", (int)(strlen(fileName) - 5), fileName);

            FILE * outputFile = fopen(outputFileName, "w");
            if (outputFile == NULL) {
                fprintf(stderr, "Error: Failed to open output file\n");
                fclose(inputFile);
                return 1;
            }

            CompilationEngine * compilationEngine = initCompilationEngine(inputFile, outputFile);
            if (compilationEngine == NULL) {
                fprintf(stderr, "Error: Failed to initialize compilation engine\n");
                fclose(inputFile);
                fclose(outputFile);
                return 1;
            }
            
            compileClass(compilationEngine);
            cleanupCompilationEngine(compilationEngine);
            fclose(inputFile);
            fclose(outputFile);
            return 0;
        } else {
            fprintf(stderr, "Error: Invalid file type\n");
        }
    } else {
        fprintf(stderr, "Error: File not found\n");
    }

    return 1;
}
